File: Frontend/Web/ground/utils/icon-parser.ts
/**
 * @file icon-parser.ts
 * @description Исправлено перекрытие заголовков и вложенность тегов для Dragonleaf.
 */

const DEFAULT_TEXT_CLASS = 'text-default';

const IMAGE_FORMATS = [
    {type: 'image/webp', ext: 'webp', path: '/images/fonticon/webp'},
    {type: 'image/avif', ext: 'avif', path: '/images/fontIcon/avif'},
];

const DEFAULT_IMAGE_FORMAT = IMAGE_FORMATS.find(f => f.ext === 'webp') || IMAGE_FORMATS[0];

const TYPE_GENERATOR_ALIASES: { [key: string]: string } = {
    'Armor': 'TypeArmor'
};

const SPECIAL_LOGIC_ICON = "Star";
const START_LINE_KEYWORDS = ['This'];
const ITEM_WORD_VARIANTS = 'items?|Items?';

const ICON_FILES = [
    "Day", "Hob", "Pet", "Buff", "Burn", "Buzz", "Coin", "Dawn", "Dorf", "Dusk", "Gold", "Life",
    "Luck", "Mana", "Miss", "Sage", "Soul", "Star", "Stun", "Tink", "Armor", "Bleed", "Blind", "Chana", "Chill",
    "Curse", "Enoch", "Haste", "Night", "Ronan", "Damage", "Debuff", "Harkon", "Health", "Morrow",
    "Pepper", "Poison", "Potion", "Resist", "Shared", "Static", "Thorns", "Celeste", "Empower", "Fatigue",
    "Stamina", "TypeBag", "TypeRat", "Accuracy", "Cooldown", "Insanity", "Resisted", "TypeFish", "TypeFood",
    "TypeTool", "Lifesteal", "MaxHealth", "Snowballs", "Stopwatch", "TypeArmor", "TypeCharm", "TypePlant",
    "TypeSkull", "CritChance", "CritDamage", "Nymphedora", "TypePotion", "MeleeWeapon", "TypeMineral",
    "RangedWeapon", "Regeneration", "StaminaUsage", "TypeAccessory", "TypeIngredient", "StaminaRecovery"
];

const ALIAS_MAP: { [key: string]: string } = {
    'Melee Weapon': 'MeleeWeapon', 'Ranged Weapon': 'RangedWeapon', 'Stamina Usage': 'StaminaUsage',
    'Stamina Recovery': 'StaminaRecovery', 'Crit Chance': 'CritChance', 'Crit Damage': 'CritDamage',
    'Max Health': 'MaxHealth', 'Accessory': 'TypeAccessory', 'Ingredient': 'TypeIngredient',
    'Mineral': 'TypeMineral', 'Potion': 'TypePotion', 'Food': 'TypeFood', 'Fish': 'TypeFish',
    'Plant': 'TypePlant', 'Skull': 'TypeSkull', 'Charm': 'TypeCharm', 'Bag': 'TypeBag',
    'Tool': 'TypeTool', 'Coin': 'Gold', 'Hob Gang': 'Hob',
    'stat_damageMin': 'Damage', 'stat_damageMax': 'Damage', 'stat_staminaCost': 'StaminaUsage',
    'stat_cooldown': 'Cooldown', 'stat_accuracy': 'Accuracy', 'stat_criticalChance': 'CritChance',
    'stat_criticalDamage': 'CritDamage', 'stat_critical': 'CritDamage', 'stat_stamina': 'Stamina'
};

const ICON_TRIGGER_EXCEPTIONS = ['Bag of Flour', 'Pet Rock'];
const HEADER_EXCEPTIONS_BEFORE_COLON = ['to', 'more', 'less', 'higher', 'lower', 'inside', 'slot', 'Стоимость', 'Cost', 'Hero', 'Герой'];
const HEADER_EXCEPTIONS_NO_COLON = ['are', 'is', 'can', 'require', 'also', 'rerolls', 'gain', 'inflict', 'steal', 'benefits', 'counts', 'per'];

const FORBIDDEN_NAMES = [
    'Gain', 'Get', 'Give', 'Turn', 'Shoot', 'When', 'After', 'Every', 'Activate', 'Use', 'Steal',
    'Have', 'Inflict', 'Remove', 'Trap', 'Cleanse', 'Become', 'Per', 'On', 'Increased', 'If',
    'Reduce', 'Heal', 'Look', 'Spend', 'Deal', 'Next', 'Also', 'Increase', 'Transform', 'Your',
    'That', 'The', 'Add', 'Transfer', 'Attract', 'Trigger', 'Both', 'Restore', 'Brings',
    'Opponent', 'Crack', 'Find', 'Catch', 'No', 'Absorb', 'Block', 'Hits', 'Hammer', 'Break',
    'Drain', 'Suffer', 'Void', 'Sabotage', "Weaken", "Turns"
];

const KEYWORDS_REGEX = /\b(Discharge|common|magnetic|Sabotage|empty|Lumps\s+of\s+Coal|Hunter['’]s\s+mark|Call\s+of\s+the\s+Void)\b/gi;
const ITEM_PREFIX_KEYWORDS = ['magnetic', 'common'];
const ITEM_SCALING_PREPOSITIONS = ['per', 'for each'];
const ITEM_SUFFIX_PREFIX_PATTERN = `(<span class="value-text">[\\d.-]+<\\/span>\\s*(?:(?:${ITEM_PREFIX_KEYWORDS.join('|')})\\s+|<span[^>]*>.*?<\\/span>\\s*)?|${ITEM_SCALING_PREPOSITIONS.join('|')}\\s+)`;

const FORCED_HEADER_KEYWORDS = ['Quantity', '• After'];
const HEADER_BLOCK_KEYWORDS = ['If'];
const NUMBER_FOLLOW_TRIGGERS = `(%|seconds?|quantity|uses?|less|simple|common|magnetic|random|free|or|${ITEM_WORD_VARIANTS}|more|by|max|min|enemy|\\[\\[ICN|[A-Z]|\\)|$)`;

// --- ВСПОМОГАТЕЛЬНЫЕ ФУНКЦИИ ---

const ALL_TRIGGER_WORDS = [...new Set([...ICON_FILES, ...Object.keys(ALIAS_MAP)])].sort((a, b) => b.length - a.length);

function createIconHtml(iconName: string, title: string): string {
    const sources = IMAGE_FORMATS.map(format =>
        `<source srcset="${format.path}/${iconName.toLowerCase()}.${format.ext}" type="${format.type}">`
    ).join('');

    return ` <picture class="text-icon" title="${title}">` +
        `${sources}` +
        `<img src="${DEFAULT_IMAGE_FORMAT.path}/${iconName.toLowerCase()}.${DEFAULT_IMAGE_FORMAT.ext}" alt="${title}" loading="lazy">` +
        `</picture> `;
}

function getAlphaId(num: number): string {
    let id = '', n = num;
    while (n >= 0) {
        id = String.fromCharCode(65 + (n % 26)) + id;
        n = Math.floor(n / 26) - 1;
    }
    return id;
}

const replaceOutsideSpans = (str: string, regex: RegExp, handler: (...args: string[]) => string) => {
    const ignoreRegex = /(<span[^>]*>.*?<\/span>|\[\[ICN[A-Z]+]]|<picture[^>]*>.*?<\/picture>|<br\s*\/?>)/g;
    const combined = new RegExp(`(${ignoreRegex.source})|${regex.source}`, 'g');
    return str.replace(combined, (match, ignoreGroup, ...rest) => {
        if (ignoreGroup) return match;
        const groups = rest.filter(v => typeof v === 'string' && v !== str);
        return handler(...groups);
    });
};

// --- ОСНОВНОЙ ПАРСЕР ---

export function parseTextWithIcons(text: string | undefined | null): string {
    if (!text) return '';

    let processedText = text.replace(/\\(([^)]+)\\)/g, (_, content: string) => content.replace(/[^a-zA-Zа-яА-Я]/g, ''));
    const iconsMap = new Map<string, string>();
    let iconCounter = 0;

    // ШАГ 1 & 2: Скрытие иконок
    let i = 0, tempText = '';
    while (i < processedText.length) {
        const remaining = processedText.substring(i);
        let matchFound = false;

        for (const word of ALL_TRIGGER_WORDS) {
            if (remaining.startsWith(word)) {
                if (ICON_TRIGGER_EXCEPTIONS.some(ex => remaining.startsWith(ex))) continue;
                const nextChar = processedText[i + word.length];
                const isConcatenated = ALL_TRIGGER_WORDS.some(w => processedText.substring(i + word.length).startsWith(w));

                if (!nextChar || !/[a-zA-Z]/.test(nextChar) || isConcatenated) {
                    const iconName = ALIAS_MAP[word] || word;
                    if (ICON_FILES.includes(iconName)) {
                        const alphaId = getAlphaId(iconCounter++);
                        const placeholder = `[[ICN${alphaId}]]`;
                        iconsMap.set(placeholder, createIconHtml(iconName, word));
                        tempText += placeholder;
                    } else tempText += word;
                    i += word.length;
                    matchFound = true;
                    break;
                }
            }
        }
        if (!matchFound) tempText += processedText[i++];
    }
    processedText = tempText;

    // ШАГ 3: Заголовки (ИСПРАВЛЕНО: возвращена проверка исключений без двоеточия)
    const forcedPart = FORCED_HEADER_KEYWORDS.map(k => k.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|');
    const blockPart = HEADER_BLOCK_KEYWORDS.join('|');

    // Регулярное выражение теперь снова опционально ищет двоеточие в конце короткой фразы
    const headerPattern = new RegExp(
        `(^|\\n|\\\\n)((?:(?:${forcedPart})[^:\\n\\\\•]+|(?![ \\t]*[•+](?![^:\\n\\\\]*:))` +
        `(?![\\s•]*\\d+%)(?!(?:\\s*${blockPart}\\b)(?![^:\\n\\\\]*:))[^:\\n\\\\•]{1,60}))(:|(?=\\n)|$)`,
        'g'
    );

    processedText = processedText.replace(headerPattern, (...args) => {
        const [start, content, end] = [args[1], args[2], args[3] || ''];
        if (!content.trim()) return args[0];

        const trimmed = content.trim();
        const endEx = new RegExp(`\\b(${HEADER_EXCEPTIONS_BEFORE_COLON.join('|')})$`, 'i');
        const noColEx = new RegExp(`\\b(${HEADER_EXCEPTIONS_NO_COLON.join('|')})\\b`, 'i');

        // Проверка исключений (TS6133 теперь решена)
        if (end === ':' && endEx.test(trimmed)) return args[0];
        if (!end && noColEx.test(content)) return args[0]; // Вот здесь используется переменная

        return `${start}<span class="value-text">${content}${end}</span>`;
    });

    // ШАГ 4 & 5: Проценты и числа
    processedText = replaceOutsideSpans(processedText, /(\(\s*[+-]\d+%.*?\))|([+-]?\d+%)/g, (m) => `<span class="value-text">${m}</span>`);
    const numRegex = new RegExp(`(?<!\\/)([+-]?[\\d.]+(?:-[\\d.]+)?|[\\d.]+)(\\s*)${NUMBER_FOLLOW_TRIGGERS}`, 'g');
    processedText = replaceOutsideSpans(processedText, numRegex, (n, s, f) => f === '%' ? `<span class="value-text">${n}%</span>${s}` : `<span class="value-text">${n}</span>${s}${f}`);

    // ШАГ 6: Названия и ключевые слова
    processedText = replaceOutsideSpans(processedText, /\b(by\s+)([\d.]+)\b/g, (b, n) => `${b}<span class="value-text">${n}</span>`);
    processedText = processedText.replace(/(^|\n|\\n)([A-Z][a-z]+\s+)([\d.]+(?:-[\d.]+)?)/g, (m, s, w, n) => n.includes('span') ? m : `${s}${w}<span class="value-text">${n}</span>`);

    const itemNamesRegex = /(?<!\()\b([A-Z][a-z]+(?:['’]s)?((\s+of)?\s+[A-Z][a-z]+(?:['’]s)?)*)\b/g;
    processedText = replaceOutsideSpans(processedText, itemNamesRegex, (m) => FORBIDDEN_NAMES.includes(m.split(/\s+/)[0].replace(/['’]s$/, "")) ? m : `<span class="value-text">${m}</span>`);

    // ШАГ 6.1: Специальная логика для Star (Placeholder Star + Name)
    iconsMap.forEach((html, placeholder) => {
        if (html.includes(`title="${SPECIAL_LOGIC_ICON}"`)) {
            const followWordRegex = new RegExp(`(${placeholder.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}[ \\xA0\\t]*)(${ITEM_WORD_VARIANTS}|[A-Z][a-z]+(?:\\s+[A-Z][a-z]+)*)`, 'g');
            processedText = processedText.replace(followWordRegex, (_match, s, name) => {
                return `${s}<span class="value-text">${name}</span>`;
            });
        }
    });

    processedText = replaceOutsideSpans(processedText, KEYWORDS_REGEX, (m) => `<span class="value-text">${m}</span>`);

    // ШАГ 6.2: Окрашивание слова item/items
    const itemSuffixRegex = new RegExp(`${ITEM_SUFFIX_PREFIX_PATTERN}(${ITEM_WORD_VARIANTS})\\b`, 'gi');
    processedText = processedText.replace(itemSuffixRegex, (_match, prefix, itemWord) => `${prefix}<span class="value-text">${itemWord}</span>`);

    const startKeywords = START_LINE_KEYWORDS.join('|');
    processedText = replaceOutsideSpans(processedText, new RegExp(`(^|\\n|\\\\n)(${startKeywords})\\b`, 'g'), (s, w) => `${s}<span class="value-text">${w}</span>`);

    // --- ФИНАЛЬНЫЕ ШАГИ ОБРАБОТКИ ТЕКСТА ---

    // 1. Возвращаем иконки
    iconsMap.forEach((html, placeholder) => {
        processedText = processedText.split(placeholder).join(html);
    });

    // 2. Переносы строк
    processedText = processedText.replace(/\\n/g, '<br>');

    // 3. ОБЕРТКА ОБЫЧНОГО ТЕКСТА
    // Исключаем буллиты и уже созданные теги
    const rawTextRegex = /([^<>\n\r•]+)(?![^<]*>)/g;
    processedText = replaceOutsideSpans(processedText, rawTextRegex, (match) => {
        const trimmed = match.trim();
        if (!trimmed) return match;

        const leadingSpace = match.match(/^\s*/)?.[0] || '';
        const trailingSpace = match.match(/\s*$/)?.[0] || '';

        return `${leadingSpace}<span class="${DEFAULT_TEXT_CLASS}">${trimmed}</span>${trailingSpace}`;
    });

    // 4. ФИКС ВЛОЖЕННОСТИ (РЕШАЕТ ПРОБЛЕМУ ЦВЕТА)
    // Если Шаг 6 или Шаг 3 создали вложенный спан, вычищаем его, оставляя приоритет за value-text
    processedText = processedText.replace(/<span class="value-text">([\s\S]*?)<span class="text-default">([\s\S]*?)<\/span>([\s\S]*?)<\/span>/g, '<span class="value-text">$1$2$3</span>');
    processedText = processedText.replace(/<span class="text-default">([\s\S]*?)<span class="value-text">([\s\S]*?)<\/span>([\s\S]*?)<\/span>/g, '<span class="text-default">$1</span><span class="value-text">$2</span><span class="text-default">$3</span>');

    // 5. Пост-клин для Dragonleaf после иконок Star
    const finalIconCleanup = new RegExp(`(title="${SPECIAL_LOGIC_ICON}".*?<\\/picture>\\s*)(${ITEM_WORD_VARIANTS}|[A-Z][a-z]+(?:\\s+[A-Z][a-z]+)*)\\b`, 'g');
    processedText = processedText.replace(finalIconCleanup, (match, icon, name) => {
        if (name.includes('class="value-text"')) return match;
        // Если имя оказалось в text-default, вынимаем его
        const cleanName = name.replace(/<[^>]*>/g, "");
        return `${icon}<span class="value-text">${cleanName}</span>`;
    });

    return processedText;
}

export function generateIconsOrText(words: string[] | undefined): string {
    if (!words || words.length === 0) return '';
    return words.map(word => {
        const t = word.trim();
        const icon = TYPE_GENERATOR_ALIASES[t] || ALIAS_MAP[t] || t;
        return ICON_FILES.includes(icon) ? createIconHtml(icon, t) : `<span class="text-fallback">${t}</span>`;
    }).join(' ');
}

File: Backend/PlayerData/builds/icon_parser.py
"""
Icon Parser Module
Синхронизировано с фронтенд-версией (icon-parser.ts).
Поддерживает генерацию HTML и структурированный парсинг для симуляции.
"""

import os
import sys
import re
import json
from pathlib import Path
from typing import Dict, List, Set, Optional, Tuple, Any, Union
from dataclasses import dataclass, field

# Добавляем корень проекта в путь для импортов
current_dir = Path(__file__).resolve().parent
project_root = current_dir.parent.parent.parent
sys.path.insert(0, str(project_root))

from sqlmodel import Session, select
from Backend.DB.database import engine
from Backend.PlayerData.models.Item import ItemDefinition

# --- КОНСТАНТЫ (Синхронизировано с TS) ---

DEFAULT_TEXT_CLASS = 'text-default'

IMAGE_FORMATS = [
    {"type": "image/webp", "ext": "webp", "path": "/images/fonticon/webp"},
    {"type": "image/avif", "ext": "avif", "path": "/images/fonticon/avif"},
]

DEFAULT_IMAGE_FORMAT = IMAGE_FORMATS[0]  # webp

TYPE_GENERATOR_ALIASES = {
    "Armor": "TypeArmor"
}

SPECIAL_LOGIC_ICON = "Star"
START_LINE_KEYWORDS = ["This"]
ITEM_WORD_VARIANTS = r"items?|Items?"

ICON_FILES = [
    "Day", "Hob", "Pet", "Buff", "Burn", "Buzz", "Coin", "Dawn", "Dorf", "Dusk", "Gold", "Life",
    "Luck", "Mana", "Miss", "Sage", "Soul", "Star", "Stun", "Tink", "Armor", "Bleed", "Blind", "Chana", "Chill",
    "Curse", "Enoch", "Haste", "Night", "Ronan", "Damage", "Debuff", "Harkon", "Health", "Morrow",
    "Pepper", "Poison", "Potion", "Resist", "Shared", "Static", "Thorns", "Celeste", "Empower", "Fatigue",
    "Stamina", "TypeBag", "TypeRat", "Accuracy", "Cooldown", "Insanity", "Resisted", "TypeFish", "TypeFood",
    "TypeTool", "Lifesteal", "MaxHealth", "Snowballs", "Stopwatch", "TypeArmor", "TypeCharm", "TypePlant",
    "TypeSkull", "CritChance", "CritDamage", "Nymphedora", "TypePotion", "MeleeWeapon", "TypeMineral",
    "RangedWeapon", "Regeneration", "StaminaUsage", "TypeAccessory", "TypeIngredient", "StaminaRecovery",
    "Fern", "Zahir", "CrashTestDucky", "Kragg"
]

# Карта алиасов
ALIAS_MAP = {
    "Crash Test Ducky": "CrashTestDucky",
    "Melee Weapon": "MeleeWeapon", "Ranged Weapon": "RangedWeapon", "Stamina Usage": "StaminaUsage",
    "Stamina Recovery": "StaminaRecovery", "Crit Chance": "CritChance", "Crit Damage": "CritDamage",
    "Max Health": "MaxHealth", "Accessory": "TypeAccessory", "Ingredient": "TypeIngredient",
    "Mineral": "TypeMineral", "Potion": "TypePotion", "Food": "TypeFood", "Fish": "TypeFish",
    "Plant": "TypePlant", "Skull": "TypeSkull", "Charm": "TypeCharm", "Bag": "TypeBag",
    "Tool": "TypeTool", "Coin": "Gold", "Hob Gang": "Hob",
    "stat_damageMin": "Damage", "stat_damageMax": "Damage", "stat_staminaCost": "StaminaUsage",
    "stat_cooldown": "Cooldown", "stat_accuracy": "Accuracy", "stat_criticalChance": "CritChance",
    "stat_criticalDamage": "CritDamage", "stat_critical": "CritDamage", "stat_stamina": "Stamina"
}

ICON_TRIGGER_EXCEPTIONS = [
    "Bag of Flour", "Pet Rock", "Brown Rat", "White Rat", "Giant Rat", 
    "Rat King", "Pack Rat", "Fire Rat", "Ice Rat", "Shadow Rat", "Rat Swarm"
]
HEADER_EXCEPTIONS_BEFORE_COLON = ["to", "more", "less", "higher", "lower", "inside", "slot", "Стоимость", "Cost", "Hero", "Герой"]
HEADER_EXCEPTIONS_NO_COLON = ["are", "is", "can", "require", "also", "rerolls", "gain", "inflict", "steal", "benefits", "counts", "per"]

FORBIDDEN_NAMES = {
    "Gain", "Get", "Give", "Turn", "Shoot", "When", "After", "Every", "Activate", "Use", "Steal",
    "Have", "Inflict", "Remove", "Trap", "Cleanse", "Become", "Per", "On", "Increased", "If",
    "Reduce", "Heal", "Look", "Spend", "Deal", "Next", "Also", "Increase", "Transform", "Your",
    "That", "The", "Add", "Transfer", "Attract", "Trigger", "Both", "Restore", "Brings",
    "Opponent", "Crack", "Find", "Catch", "No", "Absorb", "Block", "Hits", "Hammer", "Break",
    "Drain", "Suffer", "Void", "Sabotage", "Weaken", "Turns", "Can"
}

KEYWORDS_REGEX = re.compile(r"\b(Discharge|common|magnetic|Sabotage|empty|Lumps\s+of\s+Coal|Hunter['’]s\s+mark|Call\s+of\s+the\s+Void)\b", re.IGNORECASE)
ITEM_PREFIX_KEYWORDS = ["magnetic", "common"]
ITEM_SCALING_PREPOSITIONS = ["per", "for each"]
ITEM_SUFFIX_PREFIX_PATTERN = f"(<span class=\"value-text\">[\\d.-]+<\\/span>\\s*(?:(?:{'|'.join(ITEM_PREFIX_KEYWORDS)})\\s+|<span[^>]*>.*?<\\/span>\\s*)?|{'|'.join(ITEM_SCALING_PREPOSITIONS)}\\s+)"

FORCED_HEADER_KEYWORDS = ["Quantity", "• After"]
HEADER_BLOCK_KEYWORDS = ["If"]
NUMBER_FOLLOW_TRIGGERS = f"(?:%|seconds?|quantity|uses?|less|simple|common|magnetic|random|free|or|{ITEM_WORD_VARIANTS}|more|by|max|min|enemy|\\[\\[ICN|[A-Z]|\\)|$)"

# --- ВСПОМОГАТЕЛЬНЫЕ ФУНКЦИИ ---

ALL_TRIGGER_WORDS = sorted(list(set(ICON_FILES + list(ALIAS_MAP.keys()))), key=len, reverse=True)

def create_icon_html(icon_name: str, title: str) -> str:
    sources = []
    for format_info in IMAGE_FORMATS:
        sources.append(
            f'<source srcset="{format_info["path"]}/{icon_name.lower()}.{format_info["ext"]}" type="{format_info["type"]}">'
        )
    
    return (
        f' <picture class="text-icon" title="{title}">'
        f"{''.join(sources)}"
        f'<img src="{DEFAULT_IMAGE_FORMAT["path"]}/{icon_name.lower()}.{DEFAULT_IMAGE_FORMAT["ext"]}" alt="{title}" loading="lazy">'
        f'</picture> '
    )

def get_alpha_id(num: int) -> str:
    id_str = ""
    n = num
    while n >= 0:
        id_str = chr(65 + (n % 26)) + id_str
        n = n // 26 - 1
    return id_str

def replace_outside_spans(text: str, regex: Union[re.Pattern, str], handler) -> str:
    if isinstance(regex, str):
        regex = re.compile(regex)
        
    # Игнорируем существующие теги
    ignore_patterns = [
        r'<span[^>]*>.*?</span>',
        r'\[\[ICN[A-Z]+]]',
        r'<picture[^>]*>.*?</picture>',
        r'<br\s*/?>'
    ]
    ignore_regex_str = f"({'|'.join(ignore_patterns)})"
    
    # Комбинируем: (игнорируемое)|(цель)
    # Используем именованную группу для игнорирования
    combined_regex = re.compile(f"{ignore_regex_str}|({regex.pattern})", re.DOTALL if "DOTALL" in str(regex.flags) else 0)
    
    def replacer(match):
        if match.group(1): # Группа игнорирования
            return match.group(1)
        
        # Передаем все группы цели в хендлер (исключая первую группу игнорирования)
        # В Python re.sub передает объект match. Мы извлекаем нужные группы.
        target_groups = match.groups()[1:] # Группы после первой
        # Фильтруем None
        valid_groups = [g for g in target_groups if g is not None]
        
        # Если хендлер принимает match, передаем его, иначе группы
        try:
            # Пытаемся вызвать с группами
            return handler(*valid_groups)
        except TypeError:
            # Если не вышло, пробуем с одним аргументом (полное совпадение цели)
            return handler(match.group(0))

    return combined_regex.sub(replacer, text)

# --- ОСНОВНОЙ ПАРСЕР ---

@dataclass
class ParsedResult:
    html: str
    structure: Dict[str, Any] = field(default_factory=lambda: {
        "triggers": [],
        "passives": [],
        "transformations": [],
        "counters": [],
        "definitions": []
    })
    icons_used: List[str] = field(default_factory=list)

def parse_effect_details(text: str) -> Optional[Dict[str, Any]]:
    """Извлекает детали эффекта (Действие, Значения, Цели)"""
    if not text: return None
    details = {"action": None, "values": {}, "targets": [], "flags": {}, "conditions": {}, "scaling": {}, "cost": {}}

    action_patterns = [
        (r"\b(Gain)\b", "Gain"), (r"\b(Inflict)\b", "Inflict"), (r"\b(Steal)\b", "Steal"),
        (r"\b(Remove)\b", "Remove"), (r"\b(Cleanse)\b", "Cleanse"), (r"\b(Heal)\b", "Heal"),
        (r"\b(Deal)\b", "Deal"), (r"\b(Stun)\b", "Stun"), (r"\b(Use)\b", "Use"),
        (r"\b(Spend)\b", "Spend"), (r"\b(Trigger)\b", "Trigger"), (r"\b(Activate)\b", "Activate"),
        (r"\b(Reduce)\b", "Reduce"), (r"\b(Increase)\b", "Increase"), (r"\b(Drain)\b", "Drain"),
        (r"\b(Block)\b", "Block"), (r"\b(Weaken)\b", "Weaken"), (r"\b(Sabotage)\b", "Sabotage"),
        (r"\b(Brings|Catch|Find|Get)\b", "Get"), (r"\b(Break)\b", "Break")
    ]
    for p, a in action_patterns:
        if re.search(p, text, re.IGNORECASE):
            details["action"] = a
            break
            
    # Парсинг статов (упрощенно)
    stats = ["Damage", "Health", "Armor", "Mana", "Stamina", "Luck", "Poison", "Burn", "Bleed"]
    for stat in stats:
        match = re.search(rf"(\d+)\s*{stat}", text, re.IGNORECASE)
        if match: details["values"][stat.lower()] = int(match.group(1))
        
    return details if details["action"] or details["values"] else None

def parse_text_with_icons(text: Optional[str]) -> ParsedResult:
    if not text:
        return ParsedResult(html="")

    # Предварительная обработка скобок
    processed_text = re.sub(r"\\\(([^)]+)\\\)", lambda m: re.sub(r'[^a-zA-Zа-яА-Я]', '', m.group(1)), text)
    
    icons_map = {}
    icon_counter = 0
    icons_used = []

    # ШАГ 1 & 2: Скрытие иконок
    i = 0
    temp_text = ""
    while i < len(processed_text):
        remaining = processed_text[i:]
        match_found = False
        for word in ALL_TRIGGER_WORDS:
            if remaining.startswith(word):
                if any(remaining.startswith(ex) for ex in ICON_TRIGGER_EXCEPTIONS): continue
                next_char = processed_text[i + len(word)] if i + len(word) < len(processed_text) else ""
                is_concatenated = any(processed_text[i + len(word):].startswith(w) for w in ALL_TRIGGER_WORDS)
                if not next_char or not next_char.isalpha() or is_concatenated:
                    icon_name = ALIAS_MAP.get(word, word)
                    if icon_name in ICON_FILES:
                        alpha_id = get_alpha_id(icon_counter)
                        placeholder = f"[[ICN{alpha_id}]]"
                        icons_map[placeholder] = create_icon_html(icon_name, word)
                        icons_used.append(icon_name)
                        temp_text += placeholder
                        icon_counter += 1
                    else:
                        temp_text += word
                    i += len(word)
                    match_found = True
                    break
        if not match_found:
            temp_text += processed_text[i]
            i += 1
    processed_text = temp_text

    # ШАГ 3: Заголовки
    forced_part = "|".join([re.escape(k) for k in FORCED_HEADER_KEYWORDS])
    block_part = "|".join(HEADER_BLOCK_KEYWORDS)
    header_pattern = re.compile(
        rf"(^|\n|\\n)((?:(?:{forced_part})[^:\n\\•]+|(?![ \t]*[•+](?![^:\n\\]*:))"
        rf"(?![ \s•]*\d+%)(?!(?:\s*{block_part}\b)(?![^:\n\\]*:))[^:\n\\•]{{1,60}}))(:|(?=\n)|$)",
        re.MULTILINE
    )

    def header_replacer(start, content, end):
        if not content.strip(): return start + content + end
        trimmed = content.strip()
        end_ex = re.compile(rf"\b({'|'.join(HEADER_EXCEPTIONS_BEFORE_COLON)})\b$", re.IGNORECASE)
        no_col_ex = re.compile(rf"\b({'|'.join(HEADER_EXCEPTIONS_NO_COLON)})\b", re.IGNORECASE)
        if end == ":" and end_ex.search(trimmed): return start + content + end
        if not end and no_col_ex.search(content): return start + content + end
        return f"{start}<span class=\"value-text\">{content}{end}</span>"

    processed_text = header_pattern.sub(lambda m: header_replacer(m.group(1), m.group(2), m.group(3) or ""), processed_text)

    # ШАГ 4 & 5: Проценты и числа
    processed_text = replace_outside_spans(processed_text, r"(\(\s*[+-]\d+%.*?\))|([+-]?\d+%)", lambda m: f'<span class="value-text">{m}</span>')
    num_regex = rf"(?<!/)([+-]?[\d.]+(?:-[\d.]+)?|[\d.]+)(\s*){NUMBER_FOLLOW_TRIGGERS}"
    processed_text = replace_outside_spans(processed_text, num_regex, lambda n, s, f: f'<span class="value-text">{n}%</span>{s}' if f == '%' else f'<span class="value-text">{n}</span>{s}{f}')

    # ШАГ 6: Ключевые слова
    processed_text = replace_outside_spans(processed_text, r"\b(by\s+)([\d.]+)\b", lambda b, n: f'{b}<span class="value-text">{n}</span>')
    processed_text = re.sub(r"(^|\n|\\n)([A-Z][a-z]+\s+)([\d.]+(?:-[\d.]+)?)$", lambda m: f"{m.group(1)}{m.group(2)}<span class=\"value-text\">{m.group(3)}</span>" if "span" not in m.group(3) else m.group(0), processed_text, flags=re.M)

    # Имена предметов
    item_names_regex = r"(?<!\()\b([A-Z][a-z]+(?:['’]s)?((?:\s+of)?\s+[A-Z][a-z]+(?:['’]s)?){0,3})\b"
    processed_text = replace_outside_spans(processed_text, item_names_regex, lambda m: m if m.split()[0] in FORBIDDEN_NAMES else f'<span class="value-text">{m}</span>')

    # Шаг 6.1: Star Logic
    for placeholder, html in icons_map.items():
        if f'title="{SPECIAL_LOGIC_ICON}"' in html:
            follow_regex = re.compile(rf"({re.escape(placeholder)}[ \xA0\t]*)({ITEM_WORD_VARIANTS}|[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*)")
            processed_text = follow_regex.sub(lambda m: f'{m.group(1)}<span class="value-text">{m.group(2)}</span>', processed_text)

    processed_text = replace_outside_spans(processed_text, KEYWORDS_REGEX, lambda m: f'<span class="value-text">{m}</span>')

    # Шаг 6.2: Item word coloring
    item_suffix_regex = re.compile(rf"{ITEM_SUFFIX_PREFIX_PATTERN}({ITEM_WORD_VARIANTS})\b", re.IGNORECASE)
    processed_text = item_suffix_regex.sub(lambda m: f'{m.group(1)}<span class="value-text">{m.group(2)}</span>', processed_text)

    # Шаг 7: Структурированный парсинг (для симуляции)
    structure = {"triggers": [], "passives": [], "transformations": [], "counters": []}
    lines = [line.strip() for line in processed_text.replace('\u2022', '').split('\n') if line.strip()]
    current_trigger = None
    for line in lines:
        t_match = re.match(r"^(On|When|Start of|Every|Health below|Opponent health below|After|While|Per).*?:?$", line, re.I)
        if t_match:
            current_trigger = {"condition": line, "effects": []}
            structure["triggers"].append(current_trigger)
        elif current_trigger:
            current_trigger["effects"].append({"text": line, "details": parse_effect_details(line)})
        else:
            structure["passives"].append({"text": line, "details": parse_effect_details(line)})

    # Финальная сборка HTML
    for placeholder, html in icons_map.items():
        processed_text = processed_text.replace(placeholder, html)
    
    processed_text = processed_text.replace('\\n', '<br>')
    
    # Обертка обычного текста
    raw_text_regex = r"([^<>\n\r•]+)(?![^<]*>)"
    def raw_wrap(match):
        m = match.group(0)
        if not m.strip(): return m
        return f'<span class="{DEFAULT_TEXT_CLASS}">{m}</span>'
    processed_text = replace_outside_spans(processed_text, raw_text_regex, raw_wrap)

    return ParsedResult(html=processed_text, structure=structure, icons_used=icons_used)

if __name__ == "__main__":
    test = "On hit: Gain 5 Damage. Every 2 seconds: Inflict 3 Poison."
    res = parse_text_with_icons(test)
    print(res.html)
